### PR TITLE
RichText: don't set selection during selectionchange

### LINKED ISSUE #13895: RichText: Non-collapsed selection abruptly interrupted when spanning multiple nodes
When holding down the mouse cursor and selecting text in a paragraph which spans multiple child nodes, the selection is abruptly terminated.

This can happen for Shift+Enter created newlines, where the paragraph is a series of `#text`, `br`, `#text` nodes:

select

And can happen as well for text which is applied some formatting, e.g. bold:

select2

I'm not entirely sure it's related, but in the course of this I also discovered some similar selection issues with uncollapsing a selection (via Shift+Arrows), where the selection starts to expand from the right (instead of continue contracting) when reaching the end of a formatted node.

select3

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #13895. When the browser updates the selection, we are currently updating the DOM and resetting selection (as we are moving towards `RichText` becoming a controlled component). Updating the DOM is needed to add the boundary class to formatting elements when the caret moves inside the element. Setting the selection is not really needed, but I thought it'd do no harm.

But... the browser may set the selection differently than we do. E.g. it may set the selection at the end of the text node before the formatting element, where we would set it at the start of the text element inside the formatting element. This causes `isRangeEqual` to fail (the range is not equal, but visually the same), and we will reset selection during `mousedown` and `mouseup`. Selection setting by the browser will be resumed at the selection we set.

The solution is to temporarily remove selection setting until we find a way to prevent it if the selection is at the same text index.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/editor/src/components/rich-text/index.js [hunk 1]
```diff
@@ -169,13 +169,14 @@ export class RichText extends Component {
...
 
-	applyRecord( record ) {
+	applyRecord( record, { domOnly } = {} ) {
 		apply( {
 			value: record,
 			current: this.editableRef,
 			multilineTag: this.multilineTag,
 			multilineWrapperTags: this.multilineWrapperTags,
 			prepareEditableTree: this.props.prepareEditableTree,
+			__unstableDomOnly: domOnly,
```
Comment (aduth) [inline packages/editor/src/components/rich-text/index.js]: Part of me wonders if this is the sort of thing where it'd be easier to just pass through whatever additional options verbatim, i.e. `...options` from a second object argument. But I can also see the opposite argument of there being value in being explicit, especially in this case where it's something perhaps temporary.

Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: Yeah, I definitely see this as something temporary.

**Comment by iseulde** [comment]
I'm not sure if I can make a test case for this. :/

**Comment by aduth** [comment]
@iseulde The third GIF in #13895 is keyboard-only and appears like it's resolved by the changes here, in case that could work well as a base for an E2E test case.

**Comment by iseulde** [comment]
Ah, it does, thanks!

**Review by aduth — APPROVED** [review]
With a basic understanding of the underlying issue and the purpose of the selection monitoring in applying the inline boundary highlighting, I think this is a sensible change that works well in my testing. Though I'd agree as well to your further point that we should be able to have more confidence that a selection we'd assign would match a browser behavior.

**Comment by iseulde** [comment]
I tried adding the following test, but it passes in puppeteer before this change.

```js
	it( 'should not lose selection direction', async () => {
		await clickBlockAppender();
		await pressKeyWithModifier( 'primary', 'b' );
		await page.keyboard.type( '1' );
		await pressKeyWithModifier( 'primary', 'b' );
		await page.keyboard.type( '23' );
		await page.keyboard.press( 'ArrowLeft' );
		await page.keyboard.down( 'Shift' );
		await page.keyboard.press( 'ArrowLeft' );
		await page.keyboard.press( 'ArrowLeft' );
		await page.keyboard.press( 'ArrowRight' );
		await page.keyboard.press( 'ArrowRight' );
		await page.keyboard.up( 'Shift' );

		// There should be no selection. The following should not do anything.
		await pressKeyWithModifier( 'primary', 'b' );

		expect( await getEditedPostContent() ).toMatchSnapshot();
	} );
```

**Comment by iseulde** [comment]
By passing, I meant that it creates the following, as expected one would expect:

```
<!-- wp:paragraph -->
<p><strong>1</strong>23</p>
<!-- /wp:paragraph -->
```

**Comment by aduth** [comment]
For me, Cmd+B does nothing in both the case where the selection is correct (this branch) and when incorrect (on master). I'm not sure if that's a bug as well. Maybe the test should type a normal letter instead in the mean time (which would destroy content if the buggy selection behavior exists as on master, and insert a character as in this branch)?

**Comment by aduth** [comment]
Example:

```js
it( 'should not lose selection direction', async () => {
	await clickBlockAppender();
	await pressKeyWithModifier( 'primary', 'b' );
	await page.keyboard.type( '1' );
	await pressKeyWithModifier( 'primary', 'b' );
	await page.keyboard.type( '34' );
	await page.keyboard.press( 'ArrowLeft' );
	await page.keyboard.down( 'Shift' );
	await page.keyboard.press( 'ArrowLeft' );
	await page.keyboard.press( 'ArrowLeft' );
	await page.keyboard.press( 'ArrowRight' );
	await page.keyboard.press( 'ArrowRight' );
	await page.keyboard.up( 'Shift' );

	// There should be no selection. The following should insert.
	await page.keyboard.type( '2' );

	expect( await getEditedPostContent() ).toMatchSnapshot();
} );
```

**Comment by iseulde** [comment]
Added an e2e test and everything passes. Confirmed that it fails for `master`.

**Pull request merged**